### ISSUE TITLE
decide source code header style

### ISSUE THREAD

**Issue by ltalirz** [aiidateam/aiida_core#1918]
Currently, there are automatic checks on the header of source files before every release, using the [`aiida_helpers`](https://github.com/aiidateam/aiida_helpers) repository (`helpers/clean_old_headers.py`, `helpers/add_copyright_notice.py`).

With the introduction of linters, people have started to add linter directives before the header.
We should decide whether this should be allowed or not (and then adapt the `aiida_helpers`)

**Comment by ltalirz** [comment]
mentioning @sphuber @giovannipizzi @DropD

**Comment by giovannipizzi** [comment]
can linter directives be put below the header? this would simplify the job of the script

**Comment by dev-zero** [comment]
@giovannipizzi yes, they can. A standalone linter directive is always valid in the current block, starting from its position. Putting it even below the `import` is valid.
Furthermore, devs should be reminded of this fact, since there seem to be a number of directives like the following:

```python
# pylint: disable=too-few-public-methods
class Foo:
    pass

class Bar:
    def moo(self):
        pass
    def too(self):
        pass
```

which disables the warning for both classes even though the second one would be fine.
`aiida/restapi/resources.py` is such an example where guessing from the vicinity of the directive and the next line, the intention was most likely to disable it for the following class/method only.

Btw, `pylint` could even warn about useless suppression like in the Foo/Bar example, but that warning was disabled in `.pylintrc`.

**Comment by ltalirz** [comment]
Ok, I guess in this case it is decided: Linter directives *must* go below the header.
Can anyone familiar with the aiida_helpers script check whether it needs to be modified in any way (perhaps not!)?

**Comment by giovannipizzi** [comment]
I think they don't need modification for these (but a check is required after applying them in case there are directives and the script tries to put the header again).

Instead modifications are needed if what should be in the header of each file needs to change (e.g. if we want to put some `from future import ...` in all files or other similar things).
Let's discuss on Thu during the presentation with @dev-zero

**Comment by sphuber** [comment]
I prepared the release for `v1.0.0a3` yesterday and so went through the helper scripts. I have maintained the following convention for header order, which I have [documented in the wiki](https://github.com/aiidateam/aiida_core/wiki/Contributors-guidelines#coding-style):
1. Optional she-bang line e.g. `#!/usr/bin/env python`
2. Coding declaration conforming to PEP 263: e.g. `# -*- coding: utf-8 -*-`
3. A copyright header defined below
4. Linter directives e.g. `# pylint: disable=too-many-arguments`
5. Module docstrings

The helper scripts did have to be adjusted slightly, which I have done and pushed those changes. In short there were two changes:
1. the `clean_old_headers.py` only checked for the exact `# -*- coding: utf-8 -*-` string, even though PEP 263 allows more formats. The check has been updated to find all valid declarations according to PEP 263 and replace them with the preferred `# -*- coding: utf-8 -*-`
2. the `add_copyright_notice.py` failed to detect an already present notice, if it did not directly follow the coding declaration. That is to say, if someone added a pylint directive or module docstring before the copyright header, the script would add the copyright *again*. This has now been fixed.
